Fix --auth=unix when used to run a conch server on Python 3

rodrigc opened the pull request:
http://twistedmatrix.com/trac/ticket/9130

This fixes the following on Python 3, tested on Linux Fedora 24:

```
twistd -n conch -d /etc/ssh/ --auth=unix
```

rodrigc committed:
Add more tests

rodrigc committed:
Use the name of the password field in struct passwd and struct spwd

rodrigc committed:
getpwnam() and getspnam() must take str or unicode arguments for username, not bytes.

On Python 3, these functions error out if passed bytes arguments.

rodrigc committed:
crypt.crypt() must take str, not bytes, otherwise it will error out on Python 3

rodrigc committed:
Pass the encrypted password as the salt argument of crypt.crypt().

The existing code doesn't get it right if the password
was encrypted with a method besides md5 such as crypt.METHOD_SHA512.
According to: https://docs.python.org/2/library/crypt.html:
   "Since a few crypt(3) extensions allow different values, with different
    sizes in the salt, it is recommended to use the full crypted password
    as salt when checking for a password."

rodrigc committed:
Use NativeStringIO

rodrigc committed:
We do not need to export __all__ in this test file

rodrigc committed:
Clean up doc strings

rodrigc committed:
Add test

rodrigc committed:
Add newsfragment

rodrigc committed:
Properly patch spwd.getspnam() during the test.

Since we use patch(), we can get rid of the unnecessary tearDown().

rodrigc committed:
Do not use hard coded strings for comparing correct crypted values

rodrigc committed:
Test some incorrect cases

jlitzingerdev commented on src/twisted/plugins/cred_unix.py:
```diff
@@ -65,7 +68,9 @@ def checkPwd(self, pwd, username, password):
-            cryptedPass = spwd.getspnam(username)[1]
+            if not isinstance(username, StringType):
+                username = username.decode('utf-8')
+            cryptedPass = spwd.getspnam(username).sp_pwdp
```
@rodrigc Sorry for jumping in randomly here, but I'd looked at this for a moment.  Since you have a PR open I thought I'd help however I could.

Looks like the sp_pwd attribute is different between python 2/3.  `sp_pwd` vs `sp_pwdp`
https://docs.python.org/3.5/library/spwd.html
https://docs.python.org/2/library/spwd.html

Which gives me the following when I run on your branch with p2 (exception only)
```
	  File "/home/jlitzing/dev/twisted/twisted-fork/src/twisted/internet/defer.py", line 150, in maybeDeferred
	    result = f(*args, **kw)
	  File "/home/jlitzing/dev/twisted/twisted-fork/src/twisted/plugins/cred_unix.py", line 100, in requestAvatarId
	    checked = self.checkSpwd(spwd, username, password)
	  File "/home/jlitzing/dev/twisted/twisted-fork/src/twisted/plugins/cred_unix.py", line 73, in checkSpwd
	    cryptedPass = spwd.getspnam(username).sp_pwdp
	exceptions.AttributeError: 'spwd.struct_spwd' object has no attribute 'sp_pwdp'
```

rodrigc replied:
Fixed.

rodrigc committed:
Use twisted.python.compat._bytesChr.

This eliminates an error on Python 3:

   tty.tcsetattr(pty.fileno(), tty.TCSANOW, attr)
  builtins.TypeError: tcsetattr: elements of attributes must be characters or integers

rodrigc committed:
struct_pwd.sp_pwd was renamed to struct_pwd.sp_pwdp in Python 3

jlitzingerdev commented on src/twisted/cred/test/test_strcred.py:
```diff
@@ -264,10 +257,55 @@ def test_unixCheckerFailsPassword(self):
+        """
+        password = "sample password ^%$"
+
+        cryptedCorrect = crypt.crypt(password, None)
```
@rodrigc Looks like another p2/3 change.  The salt parameter isn't optional in p2.  I submitted a PR with one possible fix:
https://github.com/twisted/twisted/pull/887

rodrigc replied:
Fixed

rodrigc committed:
Fix test on older Python version where crypt.crypt() did not allow
None for salt.

rodrigc committed:
Empty string for salt is not allowed in some older Python versions

rodrigc committed:
Remove unused import

rodrigc committed:
Fix salt

rodrigc commented:
@jlitzingerdev since you expressed interest in this, can you code review this change,
as per the review process: https://twistedmatrix.com/trac/wiki/ReviewProcess#Reviewers:Howtoreviewachange ?

jlitzingerdev commented:
@rodrigc No problem.  I have to step away for a few hours but I can do that later this evening.

rodrigc committed:
Skip more tests on platforms that do not have pwd

jlitzingerdev commented on src/twisted/conch/unix.py:
```diff
@@ -29,7 +29,7 @@
 from twisted.internet.error import ProcessExitedAlready
 from twisted.python import components, log
-from twisted.python.compat import nativeString
+from twisted.python.compat import _bytesChr as chr, nativeString
```
Should the built-in chr() always have python 2 semantics (certainly yes in the existing use case) in this module?  That seems like it might cause confusion down the road and be easy to miss.  Why not just use _bytesChr() in the one location where chr() is currently used?

rodrigc replied:
There are several places in the code where replacing chr() with byteschr() has been done.
For example, in conch, it has been done in these files:

```
twisted/conch/insults/window.py
twisted/conch/ssh/connection.py
twisted/conch/ssh/session.py
twisted/conch/ssh/transport.py
twisted/conch/ssh/userauth.py
twisted/conch/telnet.py
twisted/conch/test/test_transport.py
twisted/conch/test/test_userauth.py
```

In twisted/conch/unix.py, chr() is only used on one line in that file, so I think it
is OK to replace it for this module.  The pattern has been to ``import _bytesChr as chr`` and then use ``chr()`` so I am following that pattern here.

jlitzingerdev replied:
Fair enough, I didn't search for existing instances of the same pattern (should have), so I agree.

jlitzingerdev commented on src/twisted/plugins/cred_unix.py:
```diff
@@ -65,7 +68,14 @@ def checkPwd(self, pwd, username, password):
     def checkSpwd(self, spwd, username, password):
         try:
-            cryptedPass = spwd.getspnam(username)[1]
+            if not isinstance(username, StringType):
```
This comment is preference, meaning I'm not advocating for it, but I wanted to point it out for consideration.

In nearly all cases, I dislike magic numbers for an index.  This is one of the rare cases where I actually prefer it.  My reasoning is that because the indexes are explicitly described in the standard library docs they provide a succinct way to access the encrypted password without caring about the python version.  Unless subsequent viewers of this code are familiar with getspnam() then it's likely they'll go to the std library docs anyway.  Were the code to use the old indexing I would suggest a comment describing where the index is documented.

rodrigc replied:
I'm going to disagree with you on that one.  If someone has to look into the standard library,
these fields are documented quite visibly:

https://docs.python.org/2/library/pwd.html
https://docs.python.org/2/library/spwd.html

and Python 3:

https://docs.python.org/3/library/pwd.html
https://docs.python.org/3/library/spwd.html

Having the field names spelled out is more readable than some index offset.
If someone needs to look up the field, it is quite well documented.

jlitzingerdev replied:
No worries, like I said, this was opinion on my part.

jlitzingerdev commented on src/twisted/plugins/cred_unix.py:
```diff
@@ -52,7 +53,9 @@ class UNIXChecker(object):
 
     def checkPwd(self, pwd, username, password):
         try:
```
This applies to all the methods in this class -- its a shame there are no method docstrings.  The fact that so much of Twisted is documented is one of (my opinion) its strengths.  Should lack of docsstrings live in another ticket?

rodrigc replied:
There are a few functions in this file which do not have docstrings.  I'm not going to add docstrings for all of them, but I agree with you that this could be improved, so I added docstring for `checkPwd`, `checkSpwd`, and `verifyCryptedPassword`.

Those are important functions in this file, so it doesn't hurt to document them a bit better.

jlitzingerdev commented on src/twisted/plugins/cred_unix.py:
```diff
@@ -17,22 +17,23 @@
 
 
 def verifyCryptedPassword(crypted, pw):
-    if crypted[0] == '$': # md5_crypt encrypted
```
Is there any chance this breaks MD5  encrypted passwords on python 2?  It looks like the parsing here relates to a GLibc implemenation detail (http://man7.org/linux/man-pages/man3/crypt.3.html), and the crypt module in p2 doesn't have those methods.  A secondary question is of course does it matter (security vs compatibility argument)?  I'm not overly familiar with the crypt API, so I don't have test to add right this second, but it might be worth trying.

rodrigc replied:
On my Python 3 system, I ran this program to generate an MD5 password:

```
import crypt
crypt.crypt("password", crypt.METHOD_MD5)
```

I took the string which was generated: `'$1d2n7Q0.r54s'`

and ran the following program in a Python 2.6.9 virtualenv:

```
import crypt
crypted =  "$1d2n7Q0.r54s"
crypt.crypt("password", "$1d2n7Q0.r54s") == crypted
```

and the Python 2.6.9 program returned `True`.  So this doesn't break
parsing of MD5 passwords.

rodrigc replied:
Other thing to note, is that in the Python 2 documentation:
https://docs.python.org/2/library/crypt.html

there is this:

```
Since a few crypt(3) extensions allow different values, with different sizes in the salt, it is recommended to use the full crypted password as salt when checking for a password.
```

rodrigc replied:
Also, to confirm with the implementation, this is where crypt is called in CPython 2.7:

https://github.com/python/cpython/blob/2.7/Modules/cryptmodule.c#L27

and this the implementation of crypt inside glibc:

https://sourceware.org/git/?p=glibc.git;a=blob;f=crypt/crypt-entry.c;h=2271665af2d21d2aec61ebe712ffc946c035e61e;hb=HEAD#l74

If you notice in the glibc code, the crypt() implementation parses the first
few characters of the salt, and depending on what prefix is there,
it hands it off to the code which parses MD5, SHA512, SHA256, etc.

So the change in this PR is compatible with the underlying implementation.

jlitzingerdev reviewed (commented):
This looks good, nice work!  I have two concerns -- MD5 passwords and the renaming of chr() to _bytesChr for the entire module.  The remaining comments were nits about lack of docstrings (which existed prior to this fix), and pure opinion on accessing spwd.

rodrigc committed:
Add docstring

jlitzingerdev reviewed (approved):
All of my concerns have been thoroughly addressed by Craig's comments here.  Nicely done!
